rust/storage/src/evicting.rs

```diff
@@ -0,0 +1,229 @@
+//! object stores, one for the cache and one for the backing store.  Data is written to the backing
+//! store and then to the cache.  Reads are attempted from the cache and then from the backing
+//! store.  There is no effort made to dedupe writes, because it's assumed that a single writer is
+//! working on a single block at a time.  No one should write the same block to the same location
+//! at the same time, and the cost of doing so is cache inefficiency.
+
+use std::fmt::{Debug, Display};
+use std::ops::Range;
+use std::sync::Arc;
+
+use futures::stream::StreamExt;
+use object_store::path::Path;
+use object_store::{
+    GetOptions, GetResult, ListResult, MultipartUpload, ObjectMeta, ObjectStore, PutMultipartOpts,
+    PutOptions, PutPayload, PutResult, Result,
+};
+use sync42::lru::{LeastRecentlyUsedCache, Value};
+
+use bytes::Bytes;
+use futures::stream::BoxStream;
+
+#[derive(Debug, Clone)]
+struct EvictionStub {
+    size: usize,
+}
+
+impl Value for EvictionStub {
+    fn approximate_size(&self) -> usize {
+        self.size
+    }
+}
+
+pub struct EvictingObjectStore {
+    target_disk_usage: usize,
+    object_store: Arc<dyn ObjectStore>,
+    lru: Arc<LeastRecentlyUsedCache<Path, EvictionStub>>,
+}
+
+impl EvictingObjectStore {
+    pub async fn new<O: ObjectStore>(
+        object_store: O,
+        target_disk_usage: usize,
+    ) -> Result<Self, object_store::Error> {
+        let object_store = Arc::new(object_store);
+        let lru = Arc::new(LeastRecentlyUsedCache::new(1024));
+        let mut all_objects = object_store.list(None);
+        while let Some(meta) = all_objects.next().await {
+            let meta = meta?;
+            lru.insert_no_evict(meta.location.clone(), EvictionStub { size: meta.size });
+        }
+        drop(all_objects);
+        while lru.approximate_size() > target_disk_usage {
+            let Some((key, _)) = lru.pop() else {
+                break;
+            };
+            object_store.delete(&key).await?;
+        }
+        Ok(Self {
+            target_disk_usage,
+            object_store,
+            lru,
+        })
+    }
+}
+
+impl Debug for EvictingObjectStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "EvictingObjectStore")
+    }
+}
+
+impl Display for EvictingObjectStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "EvictingObjectStore")
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for EvictingObjectStore {
+    async fn put_opts(
+        &self,
+        location: &Path,
+        payload: PutPayload,
+        opts: PutOptions,
+    ) -> Result<PutResult> {
+        self.lru.insert_no_evict(
+            location.clone(),
+            EvictionStub {
+                size: payload.content_length(),
+            },
+        );
+        while self.lru.approximate_size() > self.target_disk_usage {
```

Comment:
is it by design that the eviction library does not auto evict given a capacity i.e. why is this code pushed up here instead of being in the eviction library?

Reply:
I actually implemented it differently to start with, having a drop of an object lead to a delete.  I deemed that too fragile, as it relied upon being very implicit with drop.  This is explicit and makes clear that we pop from the LRU and then free the file that was removed from the LRU.  LRU is simply a data structure for tracking evictions.  Wiring it up to do the drop would simply move this loop to code that knows nothing about files from code that knows everything about files.